modules/auxiliary/scanner/msmail/exchange_enumerator.go

```diff
@@ -0,0 +1,543 @@
+//usr/bin/env go run "$0" "$@"; exit "$?"
+
+/*
+ OWA (Outlook Webapp) is vulnerable to time-based user enumeration attacks.
+ This module leverages all known, and even some lesser-known services exposed by default
+ Exchange installations to enumerate users. It also targets Office 365 for error-based user enumeration.
+
+ Identify Command
+
+    Used for gathering information about a host that may be pointed towards an Exchange or o365 tied domain
+    Queries for specific DNS records related to Office 365 integration
+    Attempts to extract internal domain name for onprem instance of Exchange
+    Identifies services vulnerable to time-based user enumeration for onprem Exchange
+    Lists password-sprayable services exposed for onprem Exchange host
+
+ Userenum (o365) Command
+
+    Error-based user enumeration for Office 365 integrated email addresses
+
+ */
+package main
+
+import (
+	"crypto/tls"
+	b64 "encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"metasploit/module"
+	"net"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+	"strconv"
+)
+
+func main() {
+	metadata := &module.Metadata{
+		Name:        "msmailprobe",
+		Description: "Office 365 and Exchange Enumeration",
+		Authors:     []string{"poptart", "jlarose", "Vincent Yui", "grimhacker", "Nate Power", "Nick Powers", "clee-r7"},
+		Date:        "2018-11-6",
+		Type:        "single_scanner",
+		Privileged:  false,
+		References:  []module.Reference{},
+		Options: map[string]module.Option{
+			"Command":          {Type: "string", Description: "Either 'userenum' or 'identify'", Required: true, Default: "identify"},
+			"OnPrem":           {Type: "bool", Description: "Flag to specify an On-Premise instance of Exchange", Required: false, Default: "false"},
+			"O365":             {Type: "bool", Description: "Use this flag if Exchange services are hosted by Office 365", Required: false, Default: "false"},
+			"UserName":         {Type: "string", Description: "Single user name to do identity test against", Required: false, Default: ""},
+			"UserNameFilePath": {Type: "string", Description: "Path to file containing list of users", Required: false, Default: ""},
+			"Email":            {Type: "string", Description: "Single email address to do identity test against", Required: false, Default: ""},
```

Comment:
there should be common login scanner options that you can use. Otherwise, this module will be a snowflake if you use other parameter names, making it harder to use in an automated way.

Reply:
Is there documentation on common login scanner options?

Reply:
scanners will commonly include the AuthBrute mixin, which pulls in the following parameters:

https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/auxiliary/auth_brute.rb

Look at how auxiliary/scanner/teradata/teradata_odbc_login.py was implemented and try to follow the pattern. First, observe the parameters it defines, then observe the parameters that are exposed in msfconsole when you use it. These are generated because it uses the   'single_host_login_scanner' external module template, which pulls in the above AuthBrute mixin, even though it is an external module.

Reply:
The idea behind this pattern is, your login scanner should not have to care where the source of users to try is, or where the reported goes. It just gets credentials from framework and reports them using the API. If we decided to globally change the input parameters or output parameters, doing them at the framework level is preferred to every module having its own unique method.

In the python example above, we implemented a 'login_scanner' module for python as well that implements the necessary abstraction so the actual scanner module just has to focus on delivering the results.

Reply:
Side note: To report just a username over the RPC interface you will need to add an appropriate report to https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/module/external.rb#L52